realsense_camera/src/base_nodelet.cpp

```diff
@@ -1063,155 +1063,166 @@ namespace realsense_camera
+      d2do_msg.transform.translation.x = 0;
+      d2do_msg.transform.translation.y = 0;
+      d2do_msg.transform.translation.z = 0;
+      d2do_msg.transform.rotation.x = q_d2do.getX();
+      d2do_msg.transform.rotation.y = q_d2do.getY();
+      d2do_msg.transform.rotation.z = q_d2do.getZ();
+      d2do_msg.transform.rotation.w = q_d2do.getW();
+      static_tf_broadcaster_.sendTransform(d2do_msg);
+    }
+
+    if (enable_[RS_STREAM_INFRARED])
+    {
+      // Transform color frame to infrared frame
+      c2i_msg.header.stamp = transform_ts_;
+      c2i_msg.header.frame_id = base_frame_id_;
+      c2i_msg.child_frame_id = frame_id_[RS_STREAM_INFRARED];
+      c2i_msg.transform.translation.x =  color2ir_extrinsic_.translation[2];
+      c2i_msg.transform.translation.y = -color2ir_extrinsic_.translation[0];
+      c2i_msg.transform.translation.z = -color2ir_extrinsic_.translation[1];
+      c2i_msg.transform.rotation.x = 0;
+      c2i_msg.transform.rotation.y = 0;
+      c2i_msg.transform.rotation.z = 0;
+      c2i_msg.transform.rotation.w = 1;
+      static_tf_broadcaster_.sendTransform(c2i_msg);
+
+      // Transform infrared frame to infrared optical frame
+      q_i2io.setRPY(-M_PI/2, 0.0, -M_PI/2);
+      i2io_msg.header.stamp = transform_ts_;
+      i2io_msg.header.frame_id = frame_id_[RS_STREAM_INFRARED];
+      i2io_msg.child_frame_id = optical_frame_id_[RS_STREAM_INFRARED];
+      i2io_msg.transform.translation.x = 0;
+      i2io_msg.transform.translation.y = 0;
+      i2io_msg.transform.translation.z = 0;
+      i2io_msg.transform.rotation.x = q_i2io.getX();
+      i2io_msg.transform.rotation.y = q_i2io.getY();
+      i2io_msg.transform.rotation.z = q_i2io.getZ();
+      i2io_msg.transform.rotation.w = q_i2io.getW();
+      static_tf_broadcaster_.sendTransform(i2io_msg);
+    }
   }
 
   /*
    * Publish Dynamic transforms.
    */
   void BaseNodelet::publishDynamicTransforms()
   {
+    // For single camera, we can use publishStaticTransforms() to publish static transform.
+    // However, for multiple cameras we must use publishDynamicTransforms().
+
+    // In our URDF, we provide the tf from camera_link to camera_rgb_frame. The driver
+    // considers color frame as reference frame and provides other tf publication,
+    // which are tf from color to depth, color to color_optical,  depth to depth_optical,
+    // color to ir and ir to ir_optical
     tf::Transform tr;
     tf::Quaternion q;
 
-    // The color frame is used as the base frame.
-    // Hence no additional transformation is done from base frame to color frame.
-    tr.setOrigin(tf::Vector3(0, 0, 0));
-    tr.setRotation(tf::Quaternion(0, 0, 0, 1));
-    dynamic_tf_broadcaster_.sendTransform(tf::StampedTransform(tr, transform_ts_,
-          base_frame_id_, frame_id_[RS_STREAM_COLOR]));
+    if (enable_[RS_STREAM_COLOR])
+    {
+      // Transform color frame to color optical frame
+      tr.setOrigin(tf::Vector3(0, 0, 0));
+      q.setRPY(-M_PI/2, 0.0, -M_PI/2);
```

Comment:
why are we doiing this?

Reply:
Transform color frame to color optical frame, their XYZ coordinates are different